### PR TITLE
ConnectOrFail

### PR DESCRIPTION
**Pull request by dHannasch**
This is built atop https://github.com/ray-project/ray/pull/12110. Look there first. This is incredibly hard to follow on its own, but after https://github.com/ray-project/ray/pull/12110 is settled, it will become clear that this is just moving some code around without changing anything.

## Why are these changes needed?

This makes ConnectWithRetries available on its own without necessarily quitting with a fatal error on failure to connect. (This makes it available to PingPort, for example.) (Well, really for efficiency's sake I kind of want to refactor `Connect()`, but one step at a time.)
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by dHannasch** [comment]
Several tests failed in the latest but none of the same tests failed for https://github.com/ray-project/ray/pull/13367, which is just this plus using ConnectWithRetries in PingPort, so I think all the test failures here are spurious.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```
